### ISSUE TITLE
Column start greater than line length for line X

### ISSUE THREAD

**Issue by ericcornelissen** [steelbrain/atom-linter#175]
I'm getting the error `Column start (38) greater than line length (10) for line 8`even though line 8 in that file is (full file to reproduce this error at the bottom of this issue):

>     <p>These are not the droids & you're looking for</p>

The package that causes the issue is (my own) [linter-htmllint](https://github.com/ericcornelissen/linter-htmllint), but I don't think this is a duplicate of e.g. #63 because the line length in the error is incorrect.

So I went into the source code of this package and `console.log`ed the [lineText](https://github.com/steelbrain/atom-linter/blob/d3130677b793dbe9dbd0a900169c7ab008ca5a0f/src/index.js#L31), which showed `    </div>` (the value of the next line). That value is extracted from the [text-buffer](https://github.com/atom/atom/blob/27361fd97c2e3066935e9f939f4daf65edae0205/src/text-editor.js#L942) of an Atom texteditor (as seen [here](https://github.com/steelbrain/atom-linter/blob/d3130677b793dbe9dbd0a900169c7ab008ca5a0f/src/index.js#L23)) through the [lineForRow](https://github.com/atom/text-buffer/blob/master/src/text-buffer.coffee#L677) method. That text-buffer uses a buffer from [superstring](https://github.com/atom/superstring) (as seen [here](https://github.com/atom/text-buffer/blob/b08d128dc4aa90dfcc5b0dd747656c1d19af99ad/src/text-buffer.coffee#L157)), but that is where my debugging skills stop 😅 Because I can't find the `lineForRow` method in the superstring package...

If it helps anything, I'm running windows 10 and these are my community packages:
```
├── atom-material-ui@2.1.1
├── busy-signal@1.4.3
├── click-link@1.0.11
├── docblockr@0.13.6
├── editorconfig@2.2.2
├── file-icons@2.1.16
├── intentions@1.1.5
├── language-gherkin@1.0.4
├── language-gitignore@0.3.0
├── language-htaccess@0.2.0
├── language-liquid@0.6.0
├── linter@2.2.0
├── linter-eslint@8.4.0
├── linter-htmlhint@1.4.0
├── linter-htmllint@1.0.0
├── linter-jsonlint@1.3.0
├── linter-sass-lint@1.8.3
├── linter-ui-default@1.6.10
├── minimap@4.29.7
├── one-dark-material-syntax@1.0.3
├── open-file@1.4.0
├── pdf-view@0.62.0
├── pinned-tabs@2.0.4
└── teletype@0.6.0
```

* * *

##### html file to reproduce the error
```html
<!DOCTYPE html>
<html>
  <head>
  </head>
  <body>
    <div>
      <h5>Star Wars</h5>
      <p>These are not the droids & you're looking for</p>
    </div>
  </body>
</html>
```

**Comment by Arcanemagus** [comment]
This error is thrown when the point being fed to `generateRange` doesn't actually exist in the document. In this case either the point returned from `htmllint` wasn't converted to 0-indexed before being fed to `generateRange` (which it looks like it _is_), or far more likely `htmllint` is reporting an incorrect line number.

The entire point of that check is to catch when the linter you are wrapping has a bug that needs to be fixed there. (For example this check has discovered bugs in linters like ESLint, JSHint, `htmlhint`, `flake8`, and _many_ more.)

----

There are a few different ways to handle when this finds an invalid point.

The simplest way is to simply not care about errors in points from the linter by calling `TextBuffer::clipPosition` on the (0-indexed) point from the linter so Atom will find the nearest valid point before you call `generateRange`.

The way I _prefer_ is to generate an "invalid lint message" that tells the user something bad was reported from the linter and gives them a link to file a new issue about it. This is currently best implemented [here](https://github.com/AtomLinter/linter-jshint/blob/63826776c342a9aa8cb11a07d23df4a26d8a6ece/lib/helpers.js#L206) in `linter-jshint`. I've been thinking of combining that with the previous method to still show the user _about_ where the linter _might_ have meant.

**Comment by ericcornelissen** [comment]
It seems you're right @Arcanemagus, I assumed the buffer wasn't 0-indexed because the the texteditor isn't. Thank you for the response, I'll see if I can implement an invalid lint message 👍

**Comment by Arcanemagus** [comment]
> I assumed the buffer wasn't 0-indexed because the the texteditor isn't.

_All_ [Range](https://atom.io/docs/api/v1.23.3/Range)s and [Point](https://atom.io/docs/api/v1.23.3/Point)s are 0-indexed in Atom, where are you seeing otherwise?

**Comment by ericcornelissen** [comment]
The gutter of the texteditor (visually), nothing code wise.

But what I said is not entirely true. What I meant to say is that I assumed the indices in the gutter of the texteditor are 0-indexed, in which case they would line up with the buffer as I expected. However, I didn't notice that the gutter is 1-indexed until I actually used the example code I gave (I was using a larger file when I submitted the issue).
